File: 백준/Silver/20301. 반전 요세푸스/반전 요세푸스.py
from collections import deque
N, K, M = map(int, input().split())
q = deque()
for i in range(1, N+1):
    q.append(i)

cnt = 0
while cnt < K-1:
    tmp = q.popleft()
    q.append(tmp)
    cnt += 1

cnt = 0
dir = 0
answer = []
while q:
    tmp = q.popleft()
    answer.append(tmp)
    cnt += 1
    if cnt % M == 0:
        dir += 1

    if dir % 2 == 0: # 오른쪽 방향
        idx = 0
        while idx < K-1 and q:
            tmp = q.popleft()
            q.append(tmp)
            idx += 1
    else: # 왼쪽 방향
        idx = 0
        while idx < K and q:
            tmp = q.pop()
            q.appendleft(tmp)
            idx += 1
            
for i in range(N):
    print(answer[i])

File: 백준/Silver/20301. 반전 요세푸스/반전 요세푸스.java
import java.io.*;
import java.util.StringTokenizer;
import java.util.Deque;
import java.util.LinkedList;
import java.util.ArrayList;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        Deque<Integer> nums = new LinkedList<>();
        ArrayList<Integer> answer = new ArrayList<>();
        for (int i = 1; i <= N; i++) {
            nums.add(i);
        }
        int idx = 0;
        while (idx < K-1) {
            int tmp = nums.pollFirst();
            nums.addLast(tmp);
            idx++;
        }
        int cnt = 0;
        int dir = 0;
        while (!nums.isEmpty()) {
            int tmp = nums.pollFirst();
            answer.add(tmp);
            cnt++;
            if (cnt % M == 0) {
                dir++;
            }
            if (dir % 2 == 0) { // 오른쪽 방향
                idx = 0;
                while (idx < K-1 && !nums.isEmpty()) {
                    tmp = nums.pollFirst();
                    nums.addLast(tmp);
                    idx++;
                }
            } else {
                idx = 0;
                while (idx < K && !nums.isEmpty()) {
                    tmp = nums.pollLast();
                    nums.addFirst(tmp);
                    idx++;
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < answer.size(); i++) {
            sb.append(answer.get(i)).append("\n");
        }
        System.out.print(sb);
    }
}
